### DISCUSSION TITLE
Is there an existing signed PowerShell script to set the execution policy?

### DISCUSSION THREAD

**Discussion by rhubarb-geek-nz** [PowerShell/PowerShell#19091] [General]
As a developer I have my own script

```
set-executionpolicy remotesigned
```

that I pipe into both pwsh.exe and powershell.exe in order to be able to use my own scripts.

Is there an existing version of this script available that is already signed by a recognised code signing certificate?

**Comment by mklement0** [comment]
Use the `-ExecutionPolicy` parameter in a given CLI call in order to set the effective execution policy for that call.

See:

* [`powershell.exe`](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_powershell_exe), the Windows PowerShell CLI

* [`pwsh`](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_pwsh), the PowerShell (Core) CLI

**Reply by rhubarb-geek-nz** [reply]
Right, so I still need a .bat file to wrap the call to provide the -ExecutionPolicy flag. There is no PowerShell only solution.

That can be done without the flag in a .bat file....

```
echo set-executionpolicy remotesigned | powershell.exe -nologo
```

**Reply by mklement0** [reply]
Sure it can be done that way too, but it is both more verbose than `powershell.exe -nologo -ExecutionPolicy RemoteSigned` _and_ it prevents you from passing startup commands via `-Command` / `-File`.

On a general note, piping _code_ to the PowerShell CLI exhibits undesirable pseudo-interactive behavior - see https://github.com/PowerShell/PowerShell/issues/3223

An alternative is to set the `PSExecutionPolicyPreference` environment variable _before_ calling the PowerShell CLI.
(But if you want to set this _persistently_, you may as well do a one-time call to `Set-ExecutionPolicy` with `-Scope CurrentUser` or `-Scope LocalMachine`)

**Reply by rhubarb-geek-nz** [reply]
> But if you want to set this _persistently_, you may as well do a one-time call to `Set-ExecutionPolicy` with `-Scope CurrentUser` or `-Scope LocalMachine`

Yes, that is what I am trying to achieve in my script when setting up a machine, hence why I am calling Set-ExecutionPolicy.

**Reply by mklement0** [reply]
If you want to _bootstrap_ setting a _persistent_ execution policy _from a batch file_:

```powershell
:: Use pwsh.exe for PowerShell Core.
powershell -ExecutionPolicy Bypass -Command  "Set-ExecutionPolicy -Scope CurrentUser RemoteSigned -Force"  2>NUL
```

Note the use of `2>NUL`, which silences a - highly unfortunate - error message you may encounter - see ~~https://github.com/PowerShell/PowerShell/issues/12646.~~
#12032

Note, however, that this may mask a _true_ error, such as trying to set a `-Scope LocalMachine` without running with _elevation_ or the effective execution policy being governed by GPOs.

To detect those cases, a more sophisticated `-Command` argument would be required.

---

A potential alternative is to set the `PSExecutionPolicyPreference` environment variable persistently (which you could do with `setx.exe` from a batch file), but I haven't looked into how that interacts with persistent `Set-ExecutionPolicy`-specified values.

**Reply by mklement0** [reply]
Come to think of it: you can avoid the spurious error message if you specify the _same_ execution policy both via `-ExecutionPolicy` and in the `Set-ExecutionPolicy` call (it doesn't matter which policy value you're setting, given that no _script_ is being executed; also specify  `-NoProfile` to suppress profile execution, to be safe):

```cmd
:: Use the *same* policy setting in both cases; e.g.
powershell -NoProfile -ExecutionPolicy RemoteSigned -Command  "Set-ExecutionPolicy -Scope CurrentUser RemoteSigned -Force"
```

**Reply by mklement0** [reply]
I accidentally linked to the wrong issue re the spurious error message above; here's the correct link:

* #12032

**Reply by mklement0** [reply]
I clearly missed the bootstrapping angle in your original request, but now I'm unclear on why you're trying to solve that problem with a signed script, specifically.

**Reply by rhubarb-geek-nz** [reply]
The reason is I wrote that script in 2015 and was unaware of PowerShell command line options to completely bypass the signing requirement.

**Comment by Andrew74L** [comment]
For PS5.1, I do this in a cmd script:
`reg add HKLM\SOFTWARE\Microsoft\PowerShell\1\ShellIds\Microsoft.PowerShell /v ExecutionPolicy /d RemoteSigned /f`

For PS7, I do this in a PS5.1 script:
```
$PSConfig = Get-Content "$env:ProgramFiles\PowerShell\7\powershell.config.json" | ConvertFrom-Json
$PSConfig.'Microsoft.PowerShell:ExecutionPolicy' = 'RemoteSigned'
$PSConfig | ConvertTo-Json | Out-File "$env:ProgramFiles\PowerShell\7\powershell.config.json"
```

**Reply by mklement0** [reply]
I suggest using [`Set-ExecutionPolicy`](https://learn.microsoft.com/powershell/module/microsoft.powershell.security/set-executionpolicy)

**Comment by Andrew74L** [comment]
Like @rhubarb-geek-nz, these are part of machine setup scripts. No user there to use Set-ExecutionPolicy at the command line.

**Reply by mklement0** [reply]
See [above](https://github.com/orgs/PowerShell/discussions/19091#discussioncomment-5021432)
